### PR TITLE
option for handling non-persisted node upon assignment of an association

### PR DESCRIPTION
**Pull request by dpisarewski**
Solves issue #566

### CODE REVIEW

File: spec/e2e/has_one_spec.rb [hunk 1]
```diff
@@ -28,6 +28,20 @@
       it 'raises an error when trying to create a relationship' do
         expect { unsaved_node.parent = HasOneA.create }.to raise_error(Neo4j::ActiveNode::HasN::NonPersistedNodeError)
       end
+
+      context 'with enabled auto-saving' do
+        before  { Neo4j::Config[:autosave_on_assignment] = true }
+        after   { Neo4j::Config[:autosave_on_assignment] = false }
```
Comment (cheerfulstoic) [inline spec/e2e/has_one_spec.rb]: This should probably be set to whatever `Neo4j::Config[:autosave_on_assignment]` was before it was set to `true`.  We should also probably have a spec helper for this, though you don't need to do that if you don't want to ;)

**Comment by subvertallchris** [comment]
It needs to look out for nodes that can't be saved because they're invalid. This should do the trick.

```
define_method("#{name}=") do |other_node|
  handle_non_persisted_node(other_node)
  validate_persisted_for_association!
  clear_association_cache
  association_query_proxy(name).replace_with(other_node)
end

def handle_non_persisted_node(other_node)
  return unless Neo4j::Config[:autosave_on_assignment]
  other_node.save && save
end
```

It should still do the check for non-persisted nodes. It has the added bonus of making it easy for users to monkey patch `handle_non_persisted_node` if they want special behavior.

**Comment by dpisarewski** [comment]
Thanks for the comments, guys. @cheerfulstoic which kind of spec helper do you mean? A spec helper for neo4j specs only, or kinda Rspec DSL extension which users of the gem can use?

**Comment by cheerfulstoic** [comment]
I was thinking about a neo4j(-core) gem specific helper which takes a variable name and a value and sets up `before`/`after` clauses which do what you did but making sure that the value goes back to where it started.

Something like:

```ruby
  context "something" do
    let_config(:var) { value }
  end
```

**Comment by dpisarewski** [comment]
I had to extract code from `define_has_one_methods` method because of rubocop.

```
Offenses:

lib/neo4j/active_node/has_n.rb:210:7: C: Metrics/AbcSize: Assignment Branch Condition size for define_has_one_methods is too high. [18.06/18]
      def define_has_one_methods(name)
      ^^^

134 files inspected, 1 offense detected
```

**Comment by cheerfulstoic** [comment]
Yeah, that seems fine.  It's definitely a bit tight in there ABC-wise because of all of the `define_method` calls.  There also seems to be some duplication of code which might be DRYed up at some point

**Pull request merged**